Add Integration Tests

changjian-wang opened the pull request:
This pull request introduces integration tests for both the analyzer training and person directory features, and makes improvements to the analyzer training service interface and implementation. The main changes include adding comprehensive integration tests to validate end-to-end workflows, updating the analyzer training service to return structured results, and adjusting project configuration to support testing.

**Integration Tests and Service Improvements**

*Integration Test Additions:*
- Added `AnalyzerTrainingIntegrationTest` to validate the full analyzer training workflow, including training data generation, upload verification, analyzer creation, and document analysis with output structure checks.
- Added `BuildPersonDirectoryIntegrationTest` to verify person directory creation, enrollment, identification, face management, metadata updates, and cleanup using the directory service.

*Service Interface and Implementation Updates:*
- Changed the return type of `AnalyzeDocumentWithCustomAnalyzerAsync` in both `IAnalyzerTrainingService` and its implementation from `Task` to `Task<JsonDocument>`, so it now returns structured analysis results instead of just signaling completion. [[1]](diffhunk://#diff-c0eda9f23263fe52117a400f841c44eb3100aa84a15b13b13370cebf0647e987L43-R45) [[2]](diffhunk://#diff-21be92939e3ab5b5aed70709adc3be459d5b7ff0a93a0e5ebc1a2763062560caL144-R145)
- Updated the implementation to actually return the analysis result JSON from `AnalyzeDocumentWithCustomAnalyzerAsync`.

*Project and Dependency Configuration:*
- Added a new test project file `AzureAiContentUnderstandingDotNet.Tests.csproj` with necessary dependencies for xUnit, hosting, and code coverage, and project references to all relevant feature projects.

*Minor Codebase Cleanups:*
- Removed an unused `using Microsoft.Extensions.Options;` directive from `AnalyzerTraining/Program.cs` for clarity.

These changes ensure that the analyzer training and person directory features are thoroughly tested in realistic scenarios, and that the analyzer training service provides usable output for further validation and integration.

Commit:
Update async methods to return JsonDocument results

Refactor `IAnalyzerTrainingService` and `IContentExtractionService` interfaces to change return types from `Task` to `Task<JsonDocument>`, providing structured output for analysis results.

Add new tests in `AnalyzerTrainingIntegrationTest` and `ContentExtractionIntegrationTest` to verify functionality and ensure expected JSON output.

Update `AzureAiContentUnderstandingDotNet.Tests.csproj` with necessary package and project references.

Modify `ContentExtractionService` to handle new return types and implement logic for serializing and saving JSON results.

Include the new test project in the solution file `AzureAiContentUnderstandingDotNet.sln` for better organization and clarity.

Commit:
Refactor services and tests for improved clarity

Updated `AnalyzerTrainingIntegrationTest.cs` for better clarity and functionality, focusing on the integration of the analyzer training service.

Modified method signatures in `IBuildPersonDirectoryService.cs` and `BuildPersonDirectoryService.cs` to enhance type safety, returning more specific types like `HttpResponseMessage` and `List<Person>`.

Updated `Program.cs` to align with new service method return types.

Introduced `BuildPersonDirectoryIntegrationTest.cs` to comprehensively test the `IBuildPersonDirectoryService`, covering directory creation, person enrollment, face identification, and metadata management.

Enhanced `AzureContentUnderstandingFaceClient.cs` for improved HTTP response handling and added documentation in `ServiceCollectionExtensions.cs` for clarity.

Overall, these changes enhance the clarity, functionality, and robustness of the codebase.

Commit:
Refactor test summaries and method names for clarity

Updated summaries and remarks in `AnalyzerTrainingIntegrationTest.cs` and `BuildPersonDirectoryIntegrationTest.cs` to provide clearer descriptions of test purposes and steps. Renamed test methods for consistency: `AnalyzerTrainingServiceIntegrationTest` to `RunAnalyzerTrainingIntegrationTest` and `BuildPersonDirectoryAsyncTest` to `RunBuildPersonDirectoryIntegrationTest`. Enhanced remarks to include comprehensive scenarios being validated.

Commit:
Refactor integration tests and update service interfaces

- Removed unnecessary using directive in Program.cs.
- Updated UserAgent strings in integration tests for clarity.
- Renamed test methods to follow a consistent naming pattern.
- Added ClassifierIntegrationTest class for classifier-related tests.
- Modified IClassifierService to return JsonDocument for key methods.
- Updated FieldExtractionService and ManagementService to reflect new return types.
- Enhanced type safety by changing return types in AzureContentUnderstandingClient.
- Overall improvements to code clarity, consistency, and functionality.

Commit:
Update personResponse type to handle null values

Changed the type of the `personResponse` variable in the
`UpdatePersonAsync` method from `HttpResponseMessage` to
`HttpResponseMessage?`. This allows for better handling of
scenarios where the update operation might fail and return
null, improving the robustness of the method.

Commit:
Enhance tests and add Field Extraction Pro Mode support

- Updated `AnalyzerTrainingIntegrationTest.cs` with a specific SAS URL for training data.
- Improved various test files to check for empty "warnings" in result JSON.
- Introduced `FieldExtractionProModeIntegrationTest.cs` for testing Pro Mode features.
- Modified `IFieldExtractionProModeService.cs` to return `JsonDocument` for analysis methods.
- Updated `FieldExtractionProModeService.cs` to implement new interface methods.
- Standardized user agent strings in configuration.
- Overall improvements to the testing framework and content understanding services.

Commit:
Increase timeout for document analysis polling

Extended the timeout duration for the `PollResultAsync` method from 600 seconds to 1200 seconds in the `AnalyzeDocumentWithDefinedSchemaForProModeAsync` method to allow for a longer waiting period for analysis results.

Commit:
Enhance integration tests and improve code clarity

Updated comments for better clarity and modified variable declarations in multiple integration test files. Changed `trainingDataSasUrl` to a variable in `AnalyzerTrainingIntegrationTest.cs` for environment variable support. Refined comments and improved test structure in `BuildPersonDirectoryIntegrationTest.cs`, `ClassifierIntegrationTest.cs`, `ContentExtractionIntegrationTest.cs`, `FieldExtractionIntegrationTest.cs`, `FieldExtractionProModeIntegrationTest.cs`, and `ManagementIntegrationTest.cs`. Added exception handling assertions to ensure robustness during test execution.

Commit:
Enhance integration tests with DisplayName and Trait

Updated test methods in the AzureAiContentUnderstandingDotNet.Tests namespace to include a DisplayName attribute for better clarity and a Trait attribute to categorize all tests as "Integration". This improves the readability and organization of test outputs.

Commit:
Add integration test instructions to README.md

This commit introduces a new section titled "Integration Test" in the README.md file. It outlines how to run integration tests using the `dotnet test` command and highlights the need to configure the `appsettings.json` file with the Azure AI Services endpoint prior to running the tests. The section also references additional configuration instructions for clarity.

yungshinlintw commented on AzureAiContentUnderstandingDotNet.Tests/AnalyzerTrainingIntegrationTest.cs:
```diff
@@ -0,0 +1,125 @@
+using Microsoft.Extensions.Hosting;
+using System.Text.Json;
+
+namespace AzureAiContentUnderstandingDotNet.Tests
```
nit: No need for the "DotNet" postfix in the namespace name as we are already in a .NET project.

changjian-wang replied:
Resolved.

yungshinlintw commented on AzureAiContentUnderstandingDotNet.Tests/ClassifierIntegrationTest.cs:
```diff
@@ -0,0 +1,166 @@
+                var (analyzerSchemaPath, enhancedSchemaPath) = ("./analyzer_templates/analyzer_schema.json", "./data/classifier/enhanced_schema.json");
+                var classifierId = $"classifier-sample-{Guid.NewGuid()}";
+                var classifierSchemaPath = "./data/classifier/schema.json";
+
```
Can be in the next PR: Add more verification on the result

yungshinlintw commented on AzureAiContentUnderstandingDotNet.Tests/ManagementIntegrationTest.cs:
```diff
@@ -0,0 +1,106 @@
+                Assert.True(((JsonElement)details["fieldSchema"]).TryGetProperty("fields", out var fields));
+                Assert.True(!string.IsNullOrWhiteSpace(fields.GetRawText()));
+
+                // Step 3: List all analyzers (verifies listing API, no assertion)
```
Next PR: Make sure that the created analyzer is in the list.

yungshinlintw commented on AzureAiContentUnderstandingDotNet.Tests/ManagementIntegrationTest.cs:
```diff
@@ -0,0 +1,106 @@
+                // Step 3: List all analyzers (verifies listing API, no assertion)
+                await service.ListAnalyzersAsync();
+
+                // Step 4: Delete analyzer
```
Next PR: After deletion, make sure that the created analyzer is not in the list

yungshinlintw reviewed (approved):
Approve with suggestions for additions in the next PR

Commit:
Refactor test namespace and update solution configuration

Changed the namespace for multiple test classes from
`AzureAiContentUnderstandingDotNet.Tests` to
`AzureAiContentUnderstanding.Tests`. Updated the project
entry in the solution file to reflect this change and
ensured the build configurations are correctly associated
with the renamed project.
